Core support for creating order with coupon

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #9003 
Closes: #9027

Should be reviewed together with (after) https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2728. 

Before merging this PR remember to update FluxC version after the corresponding PR is merged ⚠️ 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR aims to match iOS implementation and adds core support for creating orders with coupons.
* Added new button "+ Add coupon" redirecting to the new coupon entry form
* Updated OrderSyncStrategy classes to take coupon line changes into consideration
* Updated data model and passed couponLines argument to REST client

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
Verify it's possible to apply a coupon when creating an order. Focus on checking if the order gets properly updated after adding/removing coupon to the order.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-android/assets/4527432/8ee19dbf-b327-4e50-bb06-c5656be74ebf

### Next steps
In the next PR, we'll add a local validation of the coupon.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Add core coupon lines support to order creation

samiuelson committed:
Update sync strategy to include coupon lines changes

samiuelson committed:
Update unit tests

samiuelson committed:
Update fluxCVersion

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditFormFragment.kt:
```diff
@@ -298,8 +293,24 @@ class OrderCreateEditFormFragment :
         viewModel.event.observe(viewLifecycleOwner, ::handleViewModelEvents)
     }
 
+    private fun hideProgressBars(
```
❓ It looks like we are not actually hiding the progress bars instead, we change the state of the progress (negate) to whatever the `enabled` state is. Could we rename the function name so it is better understood?

samiuelson replied:
Good catch, I've fixed function and arg names.

AnirudhBhat commented:
Thanks for your work on this @samiuelson. I haven't reviewed the code fully yet. I was just playing around with the functionality and found some strange behaviors in the order creation screen. I've attached videos to demonstrate it.

In the first video, entering the wrong coupon code before adding any products would display "Unable to save the changes" as expected. However, this message is displayed on every screen exit from the coupons screen. Also, clicking on the "retry" action would just make the snack bar disappear without informing us what happened. 

Sometimes, the coupons screen stores the invalid coupon (the coupon that we entered) and gives us the option to remove it and sometimes it shows us as if there are no coupons stored.

https://github.com/woocommerce/woocommerce-android/assets/1331230/e9f72ca1-4f32-480a-a72f-eb0aad8db608


In the second video, modifying the product quantity displays a snack bar that says “unable to save changes” giving the user an illusion of the change in product quantity operation failed.

https://github.com/woocommerce/woocommerce-android/assets/1331230/be2d15a0-8e8f-4d98-a1c7-66b326dd6679


In the third video, at one point, it disables us from changing the product quantity and randomly adds many different products all by itself.

https://github.com/woocommerce/woocommerce-android/assets/1331230/f69a89fb-9f41-4a7f-a2be-f488252e1864

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/coupon/OrderCreateCouponEditionViewModelTest.kt:
```diff
@@ -0,0 +1,73 @@
+@OptIn(ExperimentalCoroutinesApi::class)
+class OrderCreateCouponEditionViewModelTest : BaseUnitTest() {
+    @Test
+    fun `given non empty coupon when passed to coupon edition screen then should show remove button`() {
```
❓ We could test the reverse of this as well. If an empty coupon code is passed, do not show the remove button. WDYT?

samiuelson replied:
I've added the test

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/coupon/OrderCreateCouponEditionViewModelTest.kt:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Test
+    fun `when non empty coupon code entered then should show done button`() {
```
❓ We could test the reverse of this as well. If the coupon code is empty, then disable the done button. WDYT?

samiuelson replied:
I've added the tests

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/creation/coupon/OrderCreateCouponEditionViewModelTest.kt:
```diff
@@ -0,0 +1,73 @@
+            assertEquals(OrderCreateCouponEditionViewModel.UpdateCouponCode("new_code"), it)
+        }
+    }
+}
```
❓ We could write tests to verify that the proper event is triggered when 
1. Done button is clicked
2. Coupon code is removed

WDYT?

samiuelson replied:
I've added the tests

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:
```diff
@@ -387,6 +388,10 @@ class OrderCreateEditViewModel @Inject constructor(
         triggerEvent(EditFee(orderSubtotal, currentFeeValue))
     }
 
+    fun onCouponButtonClicked() {
```
❓ What do you think about adding a test to verify proper event is triggered when `onCouponButtonClicked`?

samiuelson replied:
I've added the tests

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreateEditViewModel.kt:
```diff
@@ -620,6 +625,17 @@ class OrderCreateEditViewModel @Inject constructor(
     }
 
+    fun onCouponEntered(couponCode: String?) {
+        _orderDraft.update { draft ->
```
❓ What do you think about writing tests for this logic?

samiuelson replied:
I've added the tests

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/coupon/OrderCreateCouponEditionFragment.kt:
```diff
@@ -0,0 +1,88 @@
+                WooThemeWithBackground {
+                    OrderCreateCouponEditionScreen(
+                        state = state,
+                        onCouponCodeChanged = { viewModel.onCouponCodeChanged(it) },
```
❓ What do you think about using method references here instead of lambda? Because of the way the lambdas are constructed internally (anonymous class), there could be unnecessary recomposition.

[This](https://multithreaded.stitchfix.com/blog/2022/08/05/jetpack-compose-recomposition/) blog goes into more detail on this topic.

samiuelson replied:
Thanks for sharing the article. Updated to use refs.

samiuelson committed:
Add unit tests

samiuelson committed:
Update fluxCVersion

samiuelson committed:
Merge branch 'trunk' into add-coupon-to-order

# Conflicts:
#	build.gradle

samiuelson committed:
Convert lambdas to refs

samiuelson committed:
Update EditFocusedOrderCreateEditViewModelTest unit tests

samiuelson committed:
Disable coupon button when no products are added

samiuelson commented:
Thanks for the review @AnirudhBhat 🙇🏼 

I've added improvements that address the issues you've discovered - the coupon button is now disabled in case there are no products added to the order. This should solve the problem caused by the fact that API doesn't allow to update an order with coupon lines if there are no products in the order. Another problem I found was related to the coupon lines model we send to the endpoint in case the coupon is already added to the order and we're adding new products. In such cases, the coupon line model should contain only the coupon code (with id set to null), otherwise API returns error.

In the next PR, I'll also add a local validation of the coupon code - checking if the code entered exists in the store.

samiuelson committed:
Fix coupon line model sent to API

It turns out we shouldn't sent the whole coupon object, just the code. The id should be null.

samiuelson committed:
Update release notes

samiuelson committed:
Update release notes

samiuelson committed:
Revert removing coupon when products cleared

samiuelson committed:
Remove unused imports

AnirudhBhat commented:
@samiuelson Thank you for your excellent work here. I did notice some strange behaviors again. Is this because we are not validating the coupons? Here's the video



https://github.com/woocommerce/woocommerce-android/assets/1331230/cf51200c-0ca7-497a-88ae-32844c45e02e

AnirudhBhat reviewed (approved):
It looks like [coupon validation PR](https://github.com/woocommerce/woocommerce-android/pull/9066) fixes the above issue I mentioned. Everything else looks good to me. :shipit:
